README.md
# vaccineEarlyInvest

<!-- badges: start -->
  [![Travis build status](https://travis-ci.com/jc-castillo/vaccineEarlyInvest.svg?branch=master)](https://travis-ci.com/jc-castillo/vaccineEarlyInvest)
<!-- badges: end -->
  
This is an R package that helps determine investments to accelerate access to a Covid-19 vaccine. It provides tools to compute the optimal portfolio of investments for a country that wants to buy production capacity before the end of clinical trials. These tools follow the analysis and the model by the [AcceleratingHT](www.acceleratinght.org) team.

## Usage

The vignette `doc/countryPorfolio.html` gives examples of how to use the package to determine the optimal portfolio for a country, compute the demand curve for a country, and how to modify the default parameters to obtain the optimal portfolio under different assumptions.

Full details about the model can be found [here](www.acceleratinght.org).

NEWS.md
# vaccineEarlyInvest

### Version 0.1.0
* First release. See README.md.


tests/testthat.R
library(testthat)
library(vaccineEarlyInvest)

test_check("vaccineEarlyInvest")


tests/testthat/test-main.R
test_that('demandPriceTaker() works',{
  skip_on_cran()
  skip_on_travis()
  population <- 31.99 # In millions
  gdp_pc <- 6.71 # In thousand dollars
  frac_high_risk <- 0.131
  loss2yr <- 0.269
  parameters <- Parameters$new(population=population, gdp_pc=gdp_pc,
                               frac_high_risk=frac_high_risk, loss2yr=loss2yr)
  demand = demandPriceTaker(parameters)
  expect_equal(class(demand),'list')
  expect_length(demand,3)
  opt = demand$optimizations
  expect_true(all(opt>=0))
  expect_is(demand$allCapacities,c("matrix", "array" ))
  expect_true(all(demand$allCapacities>=0))
  demand = demandPriceTaker(population = population, gdp_pc = gdp_pc,
                            frac_high_risk = frac_high_risk, loss2yr = loss2yr)
  expect_equal(class(demand),'list')
  expect_length(demand,3)
  opt = demand$optimizations
  expect_true(all(opt>=0))
  expect_is(demand$allCapacities,c("matrix", "array" ))
  expect_true(all(demand$allCapacities>=0))
})

test_that('portfolioPriceTaker() works',{
  skip_on_cran()
  skip_on_travis()
  population <- 31.99 # In millions
  gdp_pc <- 6.71 # In thousand dollars
  frac_high_risk <- 0.131
  loss2yr <- 0.269
  parameters <- Parameters$new(population=population, gdp_pc=gdp_pc,
                               frac_high_risk=frac_high_risk, loss2yr=loss2yr)
  port = portfolioPriceTaker(parameters,price = 10)
  expect_equal(class(port),'list')
  expect_length(port,7)
  expect_true(all(port$capacities>=0))
  expect_gte(port$totCapacity,0)
  expect_gte(port$cost,0)
  expect_gte(port$expBenefits,0)
  expect_gte(port$expCapacity,0)
  port = portfolioPriceTaker(population=population, gdp_pc=gdp_pc,
                             frac_high_risk=frac_high_risk, loss2yr=loss2yr,
                             steps = c(10),
                             price = 10,return_benefit_args = TRUE)
  expect_equal(class(port),'list')
  expect_length(port,8)
  expect_true(all(port$capacities>=0))
  expect_gte(port$totCapacity,0)
  expect_gte(port$cost,0)
  expect_gte(port$expBenefits,0)
  expect_gte(port$expCapacity,0)
  benefit_args = port$benefit_args
  expect_true('price' %in% names(benefit_args))
})



tests/testthat/test-sumdist.R
#Test files for functions sumdist() and sumdistSelf() in distributions.R

test_that('Report errors when the input is not numeric',{
  dist1 = c(0.1,0.2,0.7)
  dist2 = c(0.1,'a')
  dist3 = c('a','b')
  expect_error(sumdist(dist1,dist2))
  expect_error(sumdist(dist1,dist3))
  expect_error(sumdistSelf(dist2))
  expect_error(sumdistSelf(dist3))
})

test_that('Report error if the input is not a distribution',{
  dist_neg = c(-0.5,1.5)
  dist1 = c(0.1,0.9)
  dist_no1 = c(1,2,4)
  expect_error(sumdist(dist_neg,dist1))
  expect_error(sumdist(dist1,dist_neg))
  expect_error(sumdist(dist1,dist_no1))
  expect_error(sumdist(dist_no1,dist1))
  expect_error(sumdistSelf(dist_neg))
  expect_error(sumdistSelf(dist_no1))
})

test_that('Return a discrete distribution',{
  dist1 = c(0.1,0.2,0.7)
  dist2 = c(0.2,0.8)
  expect_equal(sum(sumdist(dist1,dist2)),1)
  expect_equal(sum(sumdist(dist1,dist1)),1)
  expect_equal(sum(sumdist(dist2,dist2)),1)
  expect_equal(sum(sumdistSelf(dist1)),1)
})


R/main.R
#' Optimal portfolio for a price taker
#'
#' Computes the optimal portfolio for a price-taking country.
#' This is done by calling optimisation function [optimizeGrid()]
#' by optimising using decreasing step sizes (see `step`)
#' on benefits function [countryNetBenefits()]
#'
#' @param parameters `Parameters` object with model parameters
#' @param population Country population (in millions)
#' @param gdp_pc Country GDP per capita (in thousand $)
#' @param frac_high_risk Fraction of population that is high risk
#' @param loss2yr Cumulative percent of GDP lost because of pandemic over two years
#' @param price Market price for manufacturing capacity ($ per course / year)
#' @param steps Steps to optimize over. By default three step-sizes are used in succession:
#'              10, 1 and 0.1.
#' @param candidateFile File with candidate data
#' @param return_benefit_args if `TRUE`, then an extra object is returned, that can then be used
#'                            as input to [countryNetBenefits()]
#' @param lambda Lagrange multiplier. Should be one, unless trying to solve
#'               problem with constrained budget.
#'
#' @return List with information on optimal portfolio. Includes investment
#'  in every candidate, benefits, cost, and total capacity.
#' @export
#' @examples
#' \dontrun{
#' population <- 31.99
#' gdp_pc <- 6.71
#' frac_high_risk <- 0.131
#' loss2yr <- 0.269
#' price <- 4
#' portfolio <- portfolioPriceTaker(population=population, gdp_pc=gdp_pc,
#'                    frac_high_risk=frac_high_risk, loss2yr=loss2yr, price=price)
#' }
portfolioPriceTaker <- function(parameters=NULL, population, gdp_pc, frac_high_risk, loss2yr,
                                price, steps=c(10,1,0.1), candidateFile=NULL, lambda=1,
                                return_benefit_args = FALSE) {
  . <- Platform <- pplat <- prob <- capacity <- NULL

  if (is.null(parameters)) {
    # Create parameters from function arguments
    par <- Parameters$new(population=population, gdp_pc=gdp_pc,
                          frac_high_risk=frac_high_risk, loss2yr=loss2yr)
  } else {
    # Use parameters object passed as an argument
    par <- parameters
  }

  d <- loadData(par, candidateFile)

  d$Target <- "Other"
  d$Target[1:5]<-"Spike"
  d$Target[10:15]<-"Recombinant"

  dordered <- candidatesFung(d, par)$dordered

  dordered <- dordered[,1:11]
  dplatforms <- unique(dordered[, .(Platform, pplat)])
  setkey(dplatforms, Platform, pplat)
  dcandidate <- copy(dordered)

  # Getting information about permutations
  targets <- c("Spike","Recombinant","Other")
  probs <- c(as.numeric(par$pspike),
             as.numeric(par$precombinant),
             as.numeric(par$potherprotein))
  targetPermutations <- getTargetPermutations(targets, probs)

  # Creating objective function
  objectiveFun <- function(capacities, grid, price, par) {
    countryNetBenefits(capacities, dcandidate, targetPermutations,
                       dplatforms, grid, price, par, lambda=lambda)
  }

  # Initialize at zero
  capacities <- rep(0, nrow(dcandidate))

  # Optimize over grids with decreasing step sizes
  for (s in steps) {
    capacities <- optimizeGrid(capacities, objectiveFun, step=s, price=price, par=par,
                               name = "country net benefits")
  }

  if(return_benefit_args)
    benefit_args <- list(capacities=capacities, dcandidate=dcandidate, targetPermutations=targetPermutations,
                         dplatforms=dplatforms, grid=min(steps), price=price, par=par, lambda=lambda)

  # Computing other values
  totCapacity <- sum(capacities)
  cost <- priceTakerCost(capacities, price)
  expBenefits <-
    countryExpectedBenefits(capacities, dcandidate,
                            targetPermutations, dplatforms, par, grid=s)
  distribution <-
    countryDistribution(capacities, dcandidate,
                        targetPermutations, dplatforms, par, grid=s)
  expCapacity <- sum(distribution[, prob * capacity])

  out <- list(capacities=capacities, totCapacity=totCapacity, cost=cost,
              expBenefits=expBenefits, distribution=distribution, expCapacity=expCapacity,
              dordered=dordered)

  if(return_benefit_args)
    out$benefit_args <-  benefit_args

  return(out)
}


#' Demand curve for a price-taking country
#'
#' Builds the demand curve for a price-taking country
#'
#' @param parameters `Parameters` object with model parameters
#' @param population Country population (in millions)
#' @param gdp_pc Country GDP per capita (in thousand $)
#' @param frac_high_risk Fraction of population that is high risk
#' @param loss2yr Cumulative percent of GDP lost because of pandemic over two years
#' @param prices Vector of market price for manufacturing capacity ($ per course / year)
#' @param candidateFile File with candidate data
#' @param inisteps Step sizes to optimize over for the initial point
#' @param mainstep Step size for main optimization
#' @param candidateFile File with candidate data
#' @param verbose How much output to produce. 0 means no output, 1 means limited output, 2 means full output.
#'
#' @return List with information on demand curve. Includes a `data.table` with total demand, benefits, and cost,
#' and a matrix with demand for individual candidates at every price
#' @export
#' @examples
#' \dontrun{
#' population <- 31.99
#' gdp_pc <- 6.71
#' frac_high_risk <- 0.131
#' loss2yr <- 0.269
#' par <- Parameters$new(population=population, gdp_pc=gdp_pc,
#'                    frac_high_risk=frac_high_risk,
#'                    loss2yr=loss2yr)
#'                    demand <- demandPriceTaker(parameters=par)
#' }
demandPriceTaker <- function(parameters=NULL, population, gdp_pc, frac_high_risk, loss2yr,
                             prices=seq(100,1,-1), inisteps=c(10,1), mainstep=0.1, candidateFile=NULL, verbose=0) {

  . <- Platform <- pplat <- ind <- expBenefit <- cost <- expCapacity <- prob <- capacity <-
    successProb <- totalCapacity <- NULL

  if (is.null(parameters)) {
    # Create parameters from function arguments
    par <- Parameters$new(global=F, inputfile="Default", maxcand=30, monthben=500, popshare=population/7800,
                          gdpshare=population*gdp_pc/1e3/87.3, fracHighRisk=frac_high_risk, afterCapacity=population/7800*500,
                          counterCapacity=population/7800*500, loss2yr=loss2yr)
  } else {
    # Copy parameters object passed as an argument
    par <- parameters
  }

  d <- loadData(par, candidateFile)

  d$Target <- "Other"
  d$Target[1:5]<-"Spike"
  d$Target[10:15]<-"Recombinant"

  dordered <- candidatesFung(d, par)$dordered

  dordered <- dordered[,1:11]
  dplatforms <- unique(dordered[, .(Platform, pplat)])
  setkey(dplatforms, Platform, pplat)
  dcandidate <- copy(dordered)

  # Getting information about permutations
  targets <- c("Spike","Recombinant","Other")
  probs <- c(as.numeric(par$pspike), as.numeric(par$precombinant), as.numeric(par$potherprotein))
  targetPermutations <- getTargetPermutations(targets, probs)

  # Creating objective function
  objectiveFun <- function(capacities, grid, price, par) {
    countryNetBenefits(capacities, dcandidate, targetPermutations, dplatforms, grid, price, par, lambda=1)
  }

  # Initialize capacities at zero, initialize price
  capacities <- rep(0, nrow(dcandidate))
  price <- prices[1]

  # Optimize over grids with decreasing step sizes to get initial point
  for (s in inisteps) {
    capacities <- optimizeGrid(capacities, objectiveFun, verbose=verbose, step=s, price=price, par=par)
  }

  # Setting up objects to save infomation
  optimizations <- data.table(price=prices)
  optimizations[, ind := .I]
  setkey(optimizations, ind)
  allCapacities <- matrix(0, length(prices), length(capacities))

  # Loop over prices to get demand levels
  for (p in prices[1:length(prices)]) {
    capacities <- optimizeGrid(capacities, objectiveFun, verbose=verbose, step=mainstep, price=p, par=par)

    i <- optimizations[price==p, ind]
    allCapacities[i, ] <- capacities
    optimizations[.(i), expBenefit :=
                    countryExpectedBenefits(capacities, dcandidate, targetPermutations, dplatforms, par, grid=mainstep)]
    optimizations[.(i), cost := priceTakerCost(capacities, p)]

    distribution <- countryDistribution(capacities, dcandidate, targetPermutations, dplatforms, par, grid=mainstep)

    optimizations[.(i), expCapacity := sum(distribution[, prob * capacity])]
    optimizations[.(i), successProb := sum(distribution[capacity > 0.011, prob])]
    optimizations[.(i), totalCapacity := sum(capacities)]
  }

  return(list(optimizations=optimizations, allCapacities=allCapacities, dordered=dordered))
}


#' Optimal portfolio with limited budget
#'
#' Computes optimal portfolios for a price-taking country when there is a limited budget at
#' different levels
#'
#' @param parameters `Parameters` object with model parameters
#' @param population Country population (in millions)
#' @param gdp_pc Country GDP per capita (in thousand $)
#' @param frac_high_risk Fraction of population that is high risk
#' @param loss2yr Cumulative percent of GDP lost because of pandemic over two years
#' @param price Market price for manufacturing capacity ($ per course / year)
#' @param shadowCosts Vector with shadow cost of public funds
#' @param candidateFile File with candidate data
#' @param inisteps Step sizes to optimize over for the initial point
#' @param mainstep Step size for main optimization
#' @param candidateFile File with candidate data
#' @param verbose How much output to produce. 0 means no output, 1 means limited output, 2 means full output.
#'
#' @return List with information on demand curve. Includes a `data.table` with total demand, benefits, and cost,
#' and a matrix with demand for individual candidates at every price
#' @export
#' @examples
#' \dontrun{
#' population <- 31.99
#' gdp_pc <- 6.71
#' frac_high_risk <- 0.131
#' loss2yr <- 0.269
#' par <- Parameters$new(population=population, gdp_pc=gdp_pc,
#'                    frac_high_risk=frac_high_risk,
#'                    loss2yr=loss2yr)
#'                    demand <- demandPriceTaker(parameters=par)
#' }
limBudgetPriceTaker <- function(parameters=NULL, population, gdp_pc, frac_high_risk, loss2yr, price,
                             shadowCosts=seq(1,5,0.1), inisteps=c(10,1), mainstep=0.1, candidateFile=NULL, verbose=0) {

  . <- Platform <- pplat <- ind <- expBenefit <- cost <- expCapacity <- prob <- capacity <-
    successProb <- totalCapacity <- NULL

  if (is.null(parameters)) {
    # Create parameters from function arguments
    par <- Parameters$new(global=F, inputfile="Default", maxcand=30, monthben=500, popshare=population/7800,
                          gdpshare=population*gdp_pc/1e3/87.3, fracHighRisk=frac_high_risk, afterCapacity=population/7800*500,
                          counterCapacity=population/7800*500, loss2yr=loss2yr)
  } else {
    # Copy parameters object passed as an argument
    par <- parameters
  }

  d <- loadData(par, candidateFile)

  d$Target <- "Other"
  d$Target[1:5]<-"Spike"
  d$Target[10:15]<-"Recombinant"

  dordered <- candidatesFung(d, par)$dordered

  dordered <- dordered[,1:11]
  dplatforms <- unique(dordered[, .(Platform, pplat)])
  setkey(dplatforms, Platform, pplat)
  dcandidate <- copy(dordered)

  # Getting information about permutations
  targets <- c("Spike","Recombinant","Other")
  probs <- c(as.numeric(par$pspike), as.numeric(par$precombinant), as.numeric(par$potherprotein))
  targetPermutations <- getTargetPermutations(targets, probs)

  # Creating objective function
  objectiveFun <- function(capacities, grid, price, par, lambda) {
    countryNetBenefits(capacities, dcandidate, targetPermutations, dplatforms, grid, price, par, lambda=lambda)
  }

  # Initialize capacities at zero, initialize price
  capacities <- rep(0, nrow(dcandidate))

  # Optimize over grids with decreasing step sizes to get initial point
  sc <- shadowCosts[1]

  for (s in inisteps) {
    capacities <- optimizeGrid(capacities, objectiveFun, verbose=verbose, step=s, price=price, par=par, lambda=sc)
  }

  # Setting up objects to save infomation
  optimizations <- data.table(shadowCost=shadowCosts)
  optimizations[, ind := .I]
  setkey(optimizations, ind)
  allCapacities <- matrix(0, length(shadowCosts), length(capacities))

  # Loop over shadowCosts to get optimal portfolios
  for (sc in shadowCosts[1:length(shadowCosts)]) {
    capacities <- optimizeGrid(capacities, objectiveFun, verbose=verbose, step=mainstep, price=price,
                               par=par, lambda=sc)

    i <- optimizations[shadowCost==sc, ind]
    allCapacities[i, ] <- capacities
    optimizations[.(i), expBenefit :=
                    countryExpectedBenefits(capacities, dcandidate, targetPermutations, dplatforms, par, grid=mainstep)]
    optimizations[.(i), cost := priceTakerCost(capacities, price)]

    distribution <- countryDistribution(capacities, dcandidate, targetPermutations, dplatforms, par, grid=mainstep)

    optimizations[.(i), expCapacity := sum(distribution[, prob * capacity])]
    optimizations[.(i), successProb := sum(distribution[capacity > 0.011, prob])]
    optimizations[.(i), totalCapacity := sum(capacities)]
  }

  return(list(optimizations=optimizations, allCapacities=allCapacities, dordered=dordered))
}


R/vaccineEarlyInvest.R
.onAttach <- function(libname, pkgname) {
  packageStartupMessage("This is vaccineEarlyInvest v0.1;")
  packageStartupMessage("see ?vaccineEarlyInvest for basic information; vignette('countryPorfolio') for a tutorial")
}


#' vaccineEarlyInvest
#' 
#' This is an R package that helps determine investments to 
#' accelerate access to a Covid-19 vaccine. It provides tools 
#' to compute the optimal portfolio of investments for a country 
#' that wants to buy production capacity before the end of clinical trials. 
#' These tools follow the analysis and the model by the [AcceleratingHT](www.acceleratinght.org) team.
#' 
#' @section Usage:
#' 
#' `vignette('countryPorfolio')` gives examples of how to 
#' use the package to determine the optimal portfolio for a country, 
#' compute the demand curve for a country, and how to modify the default 
#' parameters to obtain the optimal portfolio under different assumptions.
#' 
#' Full details about the model can be found [here](www.acceleratinght.org/link_to_appendix).
#' @docType package
#' @name vaccineEarlyInvest
NULL

R/Parameters.R
library("R6")

#' Model parameters
#'
#' @description Class to handle parameters of the model for investment in early vaccine capacity
#'
#' @import data.table
#' @export
Parameters <- R6Class("Parameters", list(
  #' @field inputfile World or US input file
  inputfile="Default",
  #' @field poverall Probability that no problem at the overall level prevents vaccine feasibility
  poverall=0.9,
  #' @field pvector Probability that there's no problem at the viral vector platform level
  pvector=0.8,
  #' @field psubunit Probability that there's no problem at the protein subunit platform level
  psubunit=0.8,
  #' @field prna Probability that there's no problem at the RNA platform level
  prna=0.6,
  #' @field pdna Probability that there's no problem at the DNA platform level
  pdna=0.4,
  #' @field pattenuated Probability that there's no problem at the live attenuated platform level
  pattenuated=0.8,
  #' @field pinactivated Probability that there's no problem at the inactivated platform level
  pinactivated=0.8,
  #' @field pvlp Probability that there's no problem at the virus-like particle platform level
  pvlp=0.8,
  #' @field ppreclinical Probability that there's no problem at the candidate level when a vaccine is in preclincal trials
  ppreclinical=0.14,
  #' @field plivebac Probability that there's no problem at the live attenuated bacteria platform level
  plivebac=0,
  #' @field paapc Probability that there's no problem at the artificial antigen presenting cells platform level
  paapc=0,
  #' @field pdendritic Probability that there's no problem at the dendritic cells platform level
  pdendritic=0,
  #' @field psav Probability that there's no problem at the self-assembling vaccine platform level
  psav=0,
  #' @field punknown Probability that there's no problem for unknown platform
  punknown=0,
  #' @field prepurposed Probability that there's no problem with a repurposed candidate
  prepurposed=0,
  #' @field pphase1 Probability that there's no problem at the candidate level when a vaccine is in phase 1 trials
  pphase1=0.23,
  #' @field pphase2 Probability that there's no problem at the candidate level when a vaccine is in phase 2 trials
  pphase2=0.32,
  #' @field pphase3 Probability that there's no problem at the candidate level when a vaccine is in phase 3 trials
  pphase3=0.5,
  #' @field psubcat Probability that there's no problem at subcategory level
  psubcat=0.8,
  #' @field pspike Probability that there's no problem at the candidate level when a vaccine targets spike proten
  pspike=1.0,
  #' @field precombinant Probability that there's no problem at the candidate level when a vaccine targets recombinant proten
  precombinant=1.0,
  #' @field potherprotein Probability that there's no problem at the candidate level when a vaccine targets some other proten
  potherprotein=1.0,

  # Country / coalition specific parameters
  #' @field popshare Share of world population accounted for by country / coalition
  popshare=1,
  #' @field gdpshare Share of world GDP accounted for by country /coalition
  gdpshare=1,
  #' @field econlossratio Ratio of economic losses in the country / coalition as a percent relative to world
  econlossratio=1,
  #' @field spillovers How much do countries in the country /coalition care about spillovers
  spillovers=0,
  #' @field outlifefactor Extent to which countries care about international lives
  outlifefactor=0,
  #' @field outlivesratio Fraction of world that lives outside country /coalition
  outlivesratio=0,
  #' @field fracHighRisk Fraction of population in the country /coalition at high risk
  fracHighRisk=0.12841,

  # Parameters from model that converts production capacity into benefits
  #' @field TT Time that the vaccine is brought forward (months)
  TT=3,
  #' @field tau Period of analysis (months)
  tau=24,
  #' @field damageshare Fraction of damage accounted for by some fraction of the population
  damageshare=0.75,
  #' @field vaccshare Fraction of population that accounts for some fraction of damage
  vaccshare=1.5/7.8,
  #' @field monthben Economic benefit for the whole world per month (in billion $)
  monthben=500,
  #' @field worldmortality World mortality
  worldmortality=200000,
  #' @field mortality Monthly internal mortality
  mortality=NULL,
  #' @field statLife Value of statistical life in million $
  statLife=1.18,
  #' @field lifeExp Life expectancy
  lifeExp=71,
  #' @field yearsLost Years lost per covid death
  yearsLost=10,
  #' @field sharm Expected share of harm avoided due to an alternative drug being developed
  sharm=0.5,
  #' @field maxcand Number of candidates to consider
  maxcand=30,
  #' @field hben Monthly health benefits
  hben=NA,
  #' @field damageint Integral that measures fraction of damages
  damageint=NA,
  #' @field totmonthben Total benefit of bringing forward vaccine for one month
  totmonthben=NA,
  #' @field alpha Parameter for heterogeneity of benefits of vaccination with a smooth functional form.
  #'  a=1 corresponds to perfectly homogeneous benefits, a=0 corresponds to all benefits to the first person.
  alpha=NA,
  #' @field benefitdist String describing the form of the benefit distribution
  benefitdist="piecewiseLinearGen",
  #' @field piecewisepar Parameters of a piecewise linear benefit distribution
  piecewisepar=NA,
  #' @field fracneeded Fraction of population needed to reopen economy
  fracneeded=0.7,
  #' @field effpop Effective world population that needs to be vaccinated to
  #' get all (100 percent) of benefits
  effpop=0.7*7.8,

  # Parameters for shape of health benefits function
  #' @field simplebenefits Whether the benefit function is a simple function with one single kink
  simplebenefits=F,
  #' @field kink2loc Location of the second kink in the benefit function
  kink2loc=0.5,
  #' @field slopefactormax Maximum ratio between the benefits from vaccinating high risk people and
  #' next population group
  slopefactormax=10,
  #' @field slopefactormin Minimum ratio between the benefits from vaccinating high risk people and
  #' next population group
  slopefactormin=5,
  #' @field maxgdpratio Maximum ratio between share of GDP and share of population
  maxgdpratio=7.2, # Norway
  #' @field benefitshape Parameter that specifies the shape of the benefit function. It's linear if it
  #' is one, it's equal to health benefits if it is zero.
  benefitshape=0.5,

  # Parameters for cost of capacity
  #' @field c Cost of capacity per dose / year
  c=2,
  #' @field capacity Constraint on capacity (doses / month)
  capacity=100,
  #' @field capkink Capacity at which marginal cost has a kink (doses / month)
  capkink=200,
  #' @field capkink_nucleic Capacity at which marginal cost has a kink for RNA and DNA (doses / month)
  capkink_nucleic=NULL,
  #' @field capkink_subunits Capacity at which marginal cost has a kink for protein subunit and viral vector (doses / month)
  capkink_subunits=NULL,
  #' @field capkink_live Capacity at which marginal cost has a kink for live attenuated and deactivated (doses / month)
  capkink_live=NULL,
  #' @field mgcostelast Elasticity of marginal cost after kink
  mgcostelast=3,
  #' @field afterCapacity Worldwide capacity after the end of the program (doses / month)
  afterCapacity=500,
  #' @field counterCapacity Worldwide capacity in the counterfactual (doses / month)
  counterCapacity=500,
  #' @field pop World population population (billions)
  pop=7.8,
  #' @field cprod Production cost per vaccine ($)
  cprod=1,
  #' @field replications Number of replications for Monte Carlo
  replications=5000,
  #' @field fcapacity Fraction of capacity cost that must be borne by the firm
  fcapacity=0.15,
  #' @field fcostred Fraction of cost reduction after the second vaccine within a platform
  fcostred=0,
  #' @field outsidefactor How much more likely are firms to be successful in an outside market
  outsidefactor=2,

  # Parameters related to procurement scheme
  #' @field pearly Price of early batch of vaccines
  pearly=35,
  #' @field plate Price of late batch of vaccines
  plate=5,
  #' @field nearly Number or vaccines in early batch
  nearly=1,
  #' @field nlate Number or vaccines in late batch
  nlate=3.5,
  #' @field riskPremium Premium that must given to candidates to participate despite risk (bn $)
  riskPremium=0.65,
  #' @field overallFung Fraction of capacity cost that is fungible across platforms
  overallFung=0,
  #' @field platFung Fraction of capacity cost that is fungible within platforms
  platFung=0,#0.2,
  #' @field subcatFung Fraction of capacity cost that is fungible within subcategory
  subcatFung=0,#0.6,
  #' @field protVectorFung Fraction of capacity cost that is fungible between protein and viral vectors
  protVectorFung=0,#0.4,
  #' @field fracScrap Fraction of cost that can be saved if a candidate is not successful
  fracScrap=0.3,
  #' @field budget Program budget
  budget=100,

  #' @field capcost Total capacity cost per candidate
  capcost=NA,
  #' @field global Whether the the analysis is for the whole wolrd
  global=F,

  # Parameters related to finding firm's incentives to participate
  #' @field outcap Capacity to build in an outside market
  outcap=0.1, # Billions
  #' @field outprob Probability that copy vaccine will be successful, conditional on own success
  outprob=0.2,
  #' @field outquantity Amount of outside vaccines that will be sold at high prices
  outquantity=0.5, # Billions

  #' @description
  #' Create a new `Parameters` object.
  #' @param input Determines how to initialize object. Can be a string telling
  #' which default parameters to use. Can also be the `input` object
  #' (of class `reactivevalues`) with the inputs from a Shiny app,
  #' in which case all inputs are copied into fields.
  #' @param population Country population (in millions)
  #' @param gdp_pc Country GDP per capita (in thousand USD)
  #' @param frac_high_risk Fraction of population that is high risk
  #' @param loss2yr Cumulative percent of GDP lost because of pandemic over two years
  #' @param worldLoss2yr Cumulative percent of GDP lost worldwide because of pandemic over two years
  #' @param benefitKinks Location of the kinks in the benefit function. List with two vectors; the first
  #' vector has the fractions vaccinated, the second has the fractions of benefits obtained.
  #' @param ... Set fields at non-default values
  #' @return A new `Parameters` object.
  #' @examples
  #' par <- Parameters$new(global = TRUE)
  #'
  initialize = function(input=NULL, population=NULL, gdp_pc=NULL, frac_high_risk=NULL,
                        loss2yr=NULL, worldLoss2yr=0.138, benefitKinks=NULL, ...) {

    if (class(input) == "reactivevalues") {
      # Copy all parameters if the input comes from a shiny app interface
      nms <- names(input)
      for (nm in nms) {
        if (nm %in% names(self)) {
          if (class(self[[nm]]) != "function") {
            self[[nm]] <- input[[nm]]
          }
        }
      }
    }

    self$capkink_nucleic <- self$capkink
    self$capkink_subunits <- self$capkink
    self$capkink_live <- self$capkink

    parlist <- list(...)

    for (i in seq_len(length(parlist))) {
      nm <- names(parlist)[i]
      if (nm %in% names(self)) {
        if (class(self[[nm]]) != "function") {
          self[[nm]] <- parlist[[nm]]
        }
      }
    }

    # Setting parameters related to country
    if (!is.null(population) & !is.null(gdp_pc) & !is.null(frac_high_risk)) {
      self$popshare <- population/7800
      self$gdpshare <- population*gdp_pc/1e3/87.3
      self$fracHighRisk <- frac_high_risk
      self$afterCapacity <- population/7800*500
      self$counterCapacity <- population/7800*500
    } else if (!is.null(population) | !is.null(gdp_pc) | !is.null(frac_high_risk)) {
      stop("The arguments for population, gdp_pc, and frac_high_risk must all be provided or none of them should be provided")
    }

    if (!is.null(loss2yr)) {
      self$econlossratio <- loss2yr/worldLoss2yr
    }

    self$setDerivedParameters(benefitKinks=benefitKinks)

    for (i in seq_len(length(parlist))) {
      nm <- names(parlist)[i]
      if (nm %in% names(self)) {
        if (class(self[[nm]]) != "function") {
          self[[nm]] <- parlist[[nm]]
        }
      }
    }


  },

  #' @description
  #' Compute additional parameters that are functions of the input parameters
  #' @param benefitKinks Location of the kinks in the benefit function. List with two vectors; the first
  #' vector has the fractions vaccinated, the second has the fractions of benefits obtained.
  setDerivedParameters = function(benefitKinks=NULL) {
    if (is.null(self$mortality)) {
      self$mortality <- self$popshare * self$worldmortality
    }

    if (!self$global) {
      popratio <- self$popshare / (0.33/7.8)
      gdpratio <- self$gdpshare / (20.54/87.3)
      gdppcratio <- gdpratio / popratio
      self$statLife=7 * gdppcratio
    }

    self$capcost <- self$c * self$capacity * 12 / 1000
    self$hben <- self$mortality * self$statLife * self$yearsLost / self$lifeExp / 1000
    self$effpop <- self$fracneeded * self$pop

    if (self$global) {
      self$totmonthben <- (self$monthben + self$hben) * (1 - self$sharm)
    } else if (self$simplebenefits) {
      self$totmonthben <- (self$econlossratio * self$monthben * self$gdpshare + self$hben) * (1 - self$sharm)
      damageratio <- 0.25 / (1-0.25) * (1-0.75) / 0.75
      self$damageshare <- 1 / (1 + (1 - self$fracHighRisk) / self$fracHighRisk * damageratio)
      self$vaccshare <- self$fracHighRisk

      self$benefitdist <- "piecewiseLinearGen"

      piecewisepar <- list(vaccshares=c(self$fracHighRisk * self$popshare / self$fracneeded, self$popshare),
                           damageshares=c(self$damageshare, 1))
      self$piecewisepar <- piecewisepar
    } else {
      self$totmonthben <- (self$econlossratio * self$monthben * self$gdpshare + self$hben) * (1 - self$sharm)
      self$benefitdist <- "piecewiseLinearGen"

      if (is.null(benefitKinks)) {
        slopefactor <- self$slopefactormin + self$gdpshare / self$popshare / self$maxgdpratio *
          (self$slopefactormax - self$slopefactormin)

        denominator <- self$fracHighRisk * slopefactor + (self$kink2loc - self$fracHighRisk) +
          (self$fracneeded - self$kink2loc)*1/2

        vaccshares <- c(self$fracHighRisk * self$popshare / self$fracneeded,
                        self$kink2loc * self$popshare / self$fracneeded, self$popshare)
        healthdamageshares <- c(slopefactor * self$fracHighRisk / denominator,
                                (self$fracHighRisk * slopefactor + (self$kink2loc - self$fracHighRisk)) / denominator,
                                1)
        lineardamageshares <- c(self$fracHighRisk / self$fracneeded, self$kink2loc /self$ fracneeded, 1)
        damageshares <- self$benefitshape * lineardamageshares + (1-self$benefitshape) * healthdamageshares

        piecewisepar <- list(vaccshares=vaccshares, damageshares=damageshares)
        self$piecewisepar <- piecewisepar
      } else {
        vaccshares <- benefitKinks[[1]] * self$popshare / self$fracneeded
        damageshares <- benefitKinks[[2]]
        piecewisepar <- list(vaccshares=vaccshares, damageshares=damageshares)
        self$piecewisepar <- piecewisepar
      }
    }

    # Compute damage parameters based on damage distribution
    if (self$benefitdist=="pnorm") {
      as <- seq(0,10,0.01)
      ys <- self$vaccshare^as + (1-self$damageshare)^as
      self$alpha <- spline(ys, as, xout=1)$y
      self$damageint <- 1-gamma(1+1/self$alpha)^2/gamma((2+self$alpha)/self$alpha)
    } else if (self$benefitdist=="piecewiseLinear") {
      slope1 <- self$damageshare / self$vaccshare
      int1 <- 0
      slope2 <- (1-self$damageshare) / (1 - self$vaccshare)
      int2 <- self$damageshare - slope2 * self$vaccshare
      self$piecewisepar <- list(slope1=slope1, int1=int1, slope2=slope2, int2=int2)
    }
  }

))
